Framework/DataObjects/src/ScanningWorkspaceBuilder.cpp

```diff
@@ -0,0 +1,323 @@
+
+/**
+ * Supply a vector of vectors which contain rotations. The inner vectors should
+ *contain the rotation for each time index, the outer vector the vector for each
+ *detector.
+ *
+ * @param rotations A vector of vectors containing rotations
+ */
+void ScanningWorkspaceBuilder::setRotations(
+    const std::vector<std::vector<Kernel::Quat>> &rotations) {
+
+  if (!m_rotations.empty() || !m_instrumentAngles.empty())
+    throw std::logic_error("Can not set rotations, as rotations or instrument "
+                           "angles have already been set.");
+
+  for (const auto &vector : rotations) {
+    verifyTimeIndexSize(vector.size(), "rotations");
+  }
+  verifyDetectorSize(rotations.size(), "rotations");
+
+  m_rotations = rotations;
+}
+
+/**
+ * Set a vector of rotations corresponding to each time index. These angles
+ *rotate the detector banks around the source, setting the corresponding
+ *positions and rotations of the detectors.
+ *
+ * Here explicit assumptions are made - that the source is at (0, 0, 0), and the
+ *rotation is in the X-Z plane. This corresponds to the common case of moving
+ *detectors to increase angular coverage.
+ *
+ * @param instrumentAngles a vector of angles, the size matching the number of
+ *time indexes
+ */
+void ScanningWorkspaceBuilder::setInstrumentAngles(
+    const std::vector<double> &instrumentAngles) {
+
+  if (!m_positions.empty() || !m_rotations.empty())
+    throw std::logic_error("Can not set instrument angles, as positions and/or "
+                           "rotations have already been set.");
+
+  verifyTimeIndexSize(instrumentAngles.size(), "instrument angles");
+  m_instrumentAngles = instrumentAngles;
+}
+
+/**
+ * Set the indexing type, either to time or detector oriented indexing.
+ *
+ * @param indexingType An index type enum
+ */
+void ScanningWorkspaceBuilder::setIndexingType(
+    const IndexingType indexingType) {
+  if (m_indexingType != IndexingType::DEFAULT)
+    throw std::logic_error("Indexing type has been set already.");
+
+  m_indexingType = indexingType;
+}
+
+/**
+ * Verify everything has been set that is required and return the workspace.
+ *
+ * @return Workspace2D with the scanning information set
+ */
+MatrixWorkspace_sptr ScanningWorkspaceBuilder::buildWorkspace() {
+  validateInputs();
+
+  auto outputWorkspace = create<Workspace2D>(
+      m_instrument, m_nDetectors * m_nTimeIndexes, m_histogram);
+
+  auto &outputDetectorInfo = outputWorkspace->mutableDetectorInfo();
+  outputDetectorInfo.setScanInterval(0, m_timeRanges[0]);
+
+  buildOutputDetectorInfo(outputDetectorInfo);
+
+  if (!m_positions.empty())
+    buildPositions(outputDetectorInfo);
+
+  if (!m_rotations.empty())
+    buildRotations(outputDetectorInfo);
+
+  if (!m_instrumentAngles.empty())
+    buildInstrumentAngles(outputDetectorInfo);
+
+  switch (m_indexingType) {
+  case IndexingType::DEFAULT:
+    outputWorkspace->setIndexInfo(
+        Indexing::IndexInfo(m_nDetectors * m_nTimeIndexes));
```

Comment:
Yes, you are right, this is a bit confusing at the moment. For now I think it makes sense to ensure the number of detectors in the instrument and `nDetectors` are the same.

I think moving the instrument into the constructor might be better, then `nDetectors` does not need to be given separately.